File: android/src/main/kotlin/com/flutterplugin/log_vault/LogVaultPlugin.kt
package com.flutterplugin.log_vault

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel

/**
 * Registers the `log_vault` MethodChannel against the running Flutter engine
 * so native (Kotlin/Java) app code can forward log lines to Dart's `LogVault`
 * via [LogVaultNative] without holding its own reference to the channel or
 * `BinaryMessenger`.
 *
 * Each [LogVaultPlugin] instance is tied to one [FlutterEngine][io.flutter.embedding.engine.FlutterEngine]
 * (add-to-app / `FlutterEngineGroup` setups can run several engines, each
 * with its own plugin instance). [onDetachedFromEngine] only clears
 * [LogVaultNative]'s channel if it still points at *this* instance's
 * channel — otherwise, engine A detaching after engine B has attached
 * would silently kill engine B's still-live logging.
 */
class LogVaultPlugin : FlutterPlugin {
    private var channel: MethodChannel? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val newChannel = MethodChannel(binding.binaryMessenger, "log_vault")
        channel = newChannel
        LogVaultNative.attach(newChannel)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        LogVaultNative.detachIf(channel)
        channel = null
    }
}

File: ios/Classes/LogVaultPlugin.swift
import Flutter
import UIKit

/// Registers the `log_vault` `FlutterMethodChannel` against the running
/// Flutter engine so native (Swift/Obj-C) app code can forward log lines
/// to Dart's `LogVault` via `LogVaultNative`, without holding its own
/// reference to the channel or `FlutterBinaryMessenger`.
///
/// Publishes itself to the engine's plugin registry via
/// `registrar.publish(_:)` so the engine invokes `detachFromEngine(for:)`
/// on teardown. Per FlutterPlugin.h, that callback is ONLY delivered to
/// instances registered through `publish:` — `addMethodCallDelegate`
/// alone does not opt in to it. Without the detach callback, an
/// add-to-app / `FlutterEngineGroup` setup with multiple engines would
/// have no way to tell `LogVaultNative` "this specific engine is gone"
/// without risking clobbering a different, still-live engine.
public class LogVaultPlugin: NSObject, FlutterPlugin {
  private var channel: FlutterMethodChannel?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = LogVaultPlugin()
    let channel = FlutterMethodChannel(
      name: "log_vault",
      binaryMessenger: registrar.messenger()
    )
    instance.channel = channel
    // Retains the instance for the engine's lifetime AND opts it in to
    // detachFromEngine(for:) — see the class doc comment. This channel is
    // native -> Dart only (Dart never calls back into native on it), so
    // no addMethodCallDelegate/handle() is needed.
    registrar.publish(instance)
    LogVaultNative.attach(channel: channel)
  }

  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    LogVaultNative.detachIf(channel: channel)
    channel = nil
  }
}
